[BUG]: Crash when stopping gossip

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

The p2p node crashed with SIGSEGV and stack trace, here's the end of the logs:

```
2022-09-21 15:36:08.781367 (p2p.Koinos) [p2p/gossip.go:404] <info>: Gossiped transaction applied - ID: 0x1220cdb0d0c54be0fa9ca548e9d7794f21c559b2110fc7a05d634e4d49d14a9e26d8 from peer Qmb...
2022-09-21 15:36:08.781501 (p2p.Koinos) [p2p/gossip.go:404] <info>: Gossiped transaction applied - ID: 0x1220246530b47ddf885ae8cd66142402940d8553378d52ad49b74d9230b7ff2cf2a4 from peer Qmb...
2022-09-21 15:36:08.792164 (p2p.Koinos) [p2p/gossip.go:404] <info>: Gossiped transaction applied - ID: 0x1220a9f28b6a02237ad47bf83926edd1e35db97ea6bf644b6e7e24e675273df03101 from peer Qmb...
2022-09-21 15:36:08.794310 (p2p.Koinos) [p2p/gossip.go:404] <info>: Gossiped transaction applied - ID: 0x122069ca9766d89fb204648fac3cdf0c1e14228b4f60f9e6d321195629dbc5890a35 from peer Qmb...
2022-09-21 15:36:08.794423 (p2p.Koinos) [p2p/gossip.go:208] <info>: Stopping gossip mode
2022-09-21 15:36:08.794746 (p2p.Koinos) [p2p/gossip.go:132] <warn>: Error getting message for topic koinos.blocks: context canceled
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xa88b4d]

goroutine 1068775 [running]:
github.com/libp2p/go-libp2p-pubsub.(*Subscription).Next(0x0, {0x1543b28, 0xc00306adc0})
        /go/pkg/mod/github.com/libp2p/go-libp2p-pubsub@v0.7.1/subscription.go:27 +0x2d
github.com/koinos/koinos-p2p/internal/p2p.(*GossipManager).readMessages(0xc000330f50, {0x1543b28, 0xc00306adc0}, 0xc004edea80)
        /koinos-p2p/internal/p2p/gossip.go:130 +0x6c
created by github.com/koinos/koinos-p2p/internal/p2p.(*GossipManager).Start
        /koinos-p2p/internal/p2p/gossip.go:88 +0x1fb
```

### Expected behavior

Not crash

### Steps to reproduce

This is a race on `gm.sub()` so it may be difficult to reproduce.

### Environment

```markdown
- OS: Ubuntu 22.04
```


### Anything else?

Looking at the code and backtrace, it is clear this is a race on `gm.sub`.

Basically `Stop()` assumes the thread will immediately exit when `cancel()` is called.

However it's possible for the following sequence of events to occur:

- The context is cancelled and `gm.sub` is set to `nil`.
- The channel in `readMessages()` receives data at about the same time.
- The `select` in `readMessages()` could go into either branch.
- According to Golang docs, `select` picks a branch at random if multiple branches are ready.
- Golang happens to pick the `ch <- msg.Data` branch
- The `select` runs to completion as it's a no-op
- The `select` is the last thing in the `readMessages()` for loop, so we go back to the top of the loop
- The first thing that happens at the top of the loop is calling `gm.sub.Next()`
- This causes `SIGSEGV` because `gm.sub` is `nil`

The proper thing to do is have `Stop()` wait for the `readMessages()` goroutine to exit after calling `cancel()`, then only assign `gm.sub = nil` after `cancel()` is finished.

--- a/internal/p2p/gossip.go
+++ b/internal/p2p/gossip.go
@@ -134,9 +134,10 @@ func readMessages(ctx context.Context, ch chan<- []byte, sub *pubsub.Subscriptio
 		}
 
 		select {
-		case ch <- msg.Data:
 		case <-ctx.Done():
 			return
+		case ch <- msg.Data:
+
 		}
 	}
 }

--- a/internal/p2p/gossip.go
+++ b/internal/p2p/gossip.go
@@ -85,7 +85,7 @@ func (gm *GossipManager) Start(ctx context.Context, ch chan<- []byte) error {
 	subCtx, cancel := context.WithCancel(ctx)
 	gm.cancel = cancel
 
-	go gm.readMessages(subCtx, ch)
+	go readMessages(subCtx, ch, gm.sub, gm.topicName)
 
 	gm.Enabled = true
 
@@ -118,18 +118,18 @@ func (gm *GossipManager) PublishMessage(ctx context.Context, bytes []byte) bool 
 	return true
 }
 
-func (gm *GossipManager) readMessages(ctx context.Context, ch chan<- []byte) {
+func readMessages(ctx context.Context, ch chan<- []byte, sub *pubsub.Subscription, topicName string) {
 	//
-	// The purpose of this function is to move messages from each topic's gm.sub.Next()
+	// The purpose of this function is to move messages from each topic's sub.Next()
 	// and send them to a single channel.
 	//
 	// Note that libp2p has already used the callback passed to RegisterValidator()
 	// to validate blocks / transactions.
 	//
 	for {
-		msg, err := gm.sub.Next(ctx)
+		msg, err := sub.Next(ctx)
 		if err != nil && !errors.Is(context.DeadlineExceeded, err) {
-			log.Warnf("Error getting message for topic %s: %s", gm.topicName, err)
+			log.Warnf("Error getting message for topic %s: %s", topicName, err)
 			return
 		}
 

